src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -384,9 +387,13 @@ private void Cancel()
                 // When cancellation propagates, SendRequestBodyAsync will set _requestCompletionState to Failed
                 requestBodyCancellationSource?.Cancel();
 
-                if (sendReset)
+                // SendReset must be called under a lock and after a cancellation of requestBodyCancellationSource.
+                lock (SyncObject)
```

Comment:
I'm not really happy to lock on the same object just after leaving the above lock block, but it seems dangerous to change order of  `requestBodyCancellationSource?.Cancel()` and `SendReset` calls as well as to move `requestBodyCancellationSource?.Cancel()` under the above lock.